syncer/dml.go

```diff
@@ -217,44 +214,79 @@ func genDeleteSQLs(param *genDMLParam) ([]string, [][]string, [][]interface{}, e
 	return sqls, keys, values, nil
 }
 
-func genDeleteSQL(qualifiedName string, value []interface{}, columns []*model.ColumnInfo, indexColumns *model.IndexInfo) (string, []interface{}) {
-	whereColumns, whereValues := columns, value
-	if indexColumns != nil {
-		whereColumns, whereValues = getColumnData(columns, indexColumns, value)
+// genInsertReplace generates a DML for `INSERT INTO` or `REPLCATE INTO`.
+// the returned SQL with placeholders for `VALUES`.
+func genInsertReplace(op, table string, columns []*model.ColumnInfo) string {
+	// NOTE: use sync.Pool to hold the builder if needed later.
+	var buf strings.Builder
+	buf.Grow(256)
+	buf.WriteString(op)
+	buf.WriteString(" " + table + " (")
+	for i, column := range columns {
+		if i != len(columns)-1 {
+			buf.WriteString("`" + strings.ReplaceAll(column.Name.O, "`", "``") + "`,")
+		} else {
+			buf.WriteString("`" + strings.ReplaceAll(column.Name.O, "`", "``") + "`)")
+		}
 	}
+	buf.WriteString(" VALUES (")
 
-	where := genWhere(whereColumns, whereValues)
-	sql := fmt.Sprintf("DELETE FROM %s WHERE %s LIMIT 1;", qualifiedName, where)
-
-	return sql, whereValues
-}
-
-func indexColumnsCount(index *model.IndexInfo) int {
-	if index == nil {
-		return 0
+	// placeholders
+	for i := range columns {
+		if i != len(columns)-1 {
+			buf.WriteString("?,")
+		} else {
+			buf.WriteString("?)")
+		}
 	}
-	return len(index.Columns)
+	return buf.String()
 }
 
-func genColumnList(columns []*model.ColumnInfo) string {
+// genUpdateSQL generates a `UPDATE` SQL with `SET` and `WHERE`.
+func genUpdateSQL(table string, updateColumns, whereColumns []*model.ColumnInfo, whereValues []interface{}) string {
 	var buf strings.Builder
-	for i, column := range columns {
-		if i != 0 {
-			buf.WriteByte(',')
+	buf.Grow(2048)
+	buf.WriteString("UPDATE ")
+	buf.WriteString(table)
+	buf.WriteString(" SET ")
+
+	for i, column := range updateColumns {
+		if i == len(updateColumns)-1 {
+			fmt.Fprintf(&buf, "`%s` = ?", strings.ReplaceAll(column.Name.O, "`", "``"))
+		} else {
+			fmt.Fprintf(&buf, "`%s` = ?, ", strings.ReplaceAll(column.Name.O, "`", "``"))
 		}
-		buf.WriteByte('`')
-		buf.WriteString(strings.ReplaceAll(column.Name.O, "`", "``"))
-		buf.WriteByte('`')
 	}
+
+	buf.WriteString(" WHERE ")
+	genWhere(&buf, whereColumns, whereValues)
+	buf.WriteString(" LIMIT 1")
 	return buf.String()
 }
 
-func genColumnPlaceholders(length int) string {
-	values := make([]string, length)
-	for i := 0; i < length; i++ {
-		values[i] = "?"
+// genDeleteSQL generates a `DELETE FROM` SQL with `WHERE`.
+func genDeleteSQL(table string, value []interface{}, columns []*model.ColumnInfo, indexColumns *model.IndexInfo) (string, []interface{}) {
+	whereColumns, whereValues := columns, value
+	if indexColumns != nil {
+		whereColumns, whereValues = getColumnData(columns, indexColumns, value)
 	}
-	return strings.Join(values, ",")
+
+	var buf strings.Builder
+	buf.Grow(1024)
```

Comment:
Can we estimate the size? If it's too hard, I think current implemention is also okay.

Reply:
a little complex to estimate the size (it may make the performance worse)